Contract to runtime call permissioning

roy-powerfinance opened the pull request:
Added delegated runtime dispatch that supports doughnut

roy-powerfinance committed:
Added delegated runtime dispatch that supports doughnut

roy-powerfinance committed:
added fork logic for when a doughnut is present when processing a contract

cowboy-bebug committed:
add a test for delegated dispatch call
- use test setup from doughnut_integration.rs, instead of tests.rs
- add delegated_dispatches field in MockExt and test whether it gets
  pushed with dispatch calls if doughnut provided

jordy25519 commented on frame/contracts/src/wasm/runtime.rs:
```diff
@@ -657,7 +657,14 @@ define_env!(Env, <E: Ext>,
 		charge_gas(&mut ctx.gas_meter, ctx.schedule, RuntimeToken::ComputedDispatchFee(fee))?;
 
-		ctx.ext.note_dispatch_call(call);
+		match ctx.ext.doughnut() {
```
There should be both a `dispatch_call` and a `dispatch_delegated_call`.
The first means contract wants to do something for itself, the second means contract wants to do something for user (if doughnut permits).
At the moment user can change contract intended behaviour by passing a doughnut

cowboy-bebug replied:
Thanks for spotting this!

cowboy-bebug replied:
Are we introducing duplicated calls by doing this?

jordy25519 reviewed (changes requested):
👍 Needs refactor so that there are two types of dispatches

cowboy-bebug committed:
add normal dispatch method call, if doughnut present
- we don't want to allow users to change contract by attaching doughnut

cowboy-bebug committed:
add a separate CODE_DELEGATED_DISPATCH_CALL to test against

cowboy-bebug committed:
add ext_delegated_dispatch_call for delegated calls with doughnut

cowboy-bebug commented on frame/contracts/src/wasm/runtime.rs:
```diff
@@ -662,6 +662,28 @@ define_env!(Env, <E: Ext>,
+
+		if let Some(doughnut) = ctx.ext.doughnut() {
+			// Use of intermediate variable to avoid `#[warn(mutable_borrow_reservation_conflict)]`
+			let d = (*doughnut).clone();
```
I don't think we need to verify doughnut at this stage (or do we?) because it's verified at
- `ext_call()` (for contract-contract calls) and
- `Contract::call()` (for runtime-contract calls)

jordy25519 replied:
Correct, no verification to do here. This is a `contract-runtime` call, it will be seen like any ordinary extrinsic with a doughnut and will be verified in the targeted runtime method

jordy25519 commented on frame/contracts/src/wasm/runtime.rs:
```diff
@@ -662,6 +662,28 @@ define_env!(Env, <E: Ext>,
 		Ok(())
 	},
 
+	ext_delegated_dispatch_call(ctx, call_ptr: u32, call_len: u32) => {
```
This is it 👍. Could use a doc comment

jordy25519 reviewed (commented):
Add an integration test and then we're good 🍩

cowboy-bebug committed:
add comments for ext_delegated_dispatch_call
- doc comments are not allowed in define_env! macro

cowboy-bebug committed:
add a test for delegated dispatch call
- alice instantiates a contract
- bob is created as a contract account
- django calls the contract (bob)
- credit is transferred from django -> charlie

cowboy-bebug committed:
add a test for delegated call with nonverifiable doughnut

cowboy-bebug commented:
~I can't trigger `verify_contract_to_contract_call` to return error anymore because `Contract::call` checks for `verify_runtime_to_contract` first before actual contract call 😞~

^ The below commit should work around this 😄

cowboy-bebug committed:
add another field to MockDoughnut to test for different verification scenarios

jordy25519 commented on frame/contracts/src/doughnut_integration.rs:
```diff
@@ -85,12 +86,17 @@ pub struct Test;
 #[derive(Clone, Eq, PartialEq, Debug, Encode, Decode)]
 pub struct MockDoughnut {
-	verifiable: bool,
+	runtime_verifiable: bool,
```
Can doc comment these for the next guy 😉

jordy25519 commented on frame/contracts/src/doughnut_integration.rs:
```diff
@@ -441,23 +454,23 @@ const CODE_CALLER_CONTRACT: &str = r#"
 	let (callee_wasm, callee_code_hash) = compile_module::<Test>(CODE_RETURN_WITH_DATA).unwrap();
 	let (caller_wasm, caller_code_hash) = compile_module::<Test>(CODE_CALLER_CONTRACT).unwrap();
-	let verifiable_doughnut = MockDoughnut::new(true);
+	let verifiable_doughnut = MockDoughnut::new(true, true);
```
Could builder pattern this but not important

cowboy-bebug replied:
Thanks for this comment! It's very clear how it's constructed now! should follow this pattern as much as possible 😄

jordy25519 commented on frame/contracts/src/doughnut_integration.rs:
```diff
@@ -521,7 +537,163 @@ fn contract_to_contract_call_returns_error_with_unverifiable_doughnut() {
+}
+
+#[test]
+fn delegated_contract_to_runtime_call_returns_error_with_unverifiable_doughnut() {
```
Name of this test is wrong as this is a test for `verify_runtime_to_contract_call`

cowboy-bebug replied:
renaming to `delegated_runtime_to_contract_call_returns_error_with_unverifiable_doughnut`

jordy25519 commented on frame/contracts/src/wasm/mod.rs:
```diff
@@ -1190,6 +1211,44 @@ mod tests {
+"#;
+
+	#[test]
+	fn delegated_dispatch_call_works_with_doughnut() {
```
Can add another test for when the doughnut is `None` here, it should error

cowboy-bebug replied:
`delegated_dispatch_call_fails_without_doughnut` is added

jordy25519 commented on frame/contracts/src/doughnut_integration.rs:
```diff
@@ -521,7 +537,163 @@ fn contract_to_contract_call_returns_error_with_unverifiable_doughnut() {
+		Balances::deposit_creating(&ALICE, 1_000_000);
+		Balances::deposit_creating(&DJANGO, 1_000_000);
+
+		assert_ok!(Contract::put_code(Origin::signed(ALICE), 100_000, wasm));
```
I notice some other tests are using `delegated_origin` to `put_code` can you fix them to use Alice like this one (I guess it was left over from when instantiate needed a `Option<Doughnut>`

jordy25519 reviewed (changes requested):
Few clean ups, otherwise keen to merge this ASAP

cowboy-bebug committed:
implement builder pattern for MockDoughnut constructor

cowboy-bebug committed:
add a test for delegated dispatch call without doughnut

cowboy-bebug committed:
rename a test, and pass in correct origin type

jordy25519 commented on frame/contracts/src/doughnut_integration.rs:
```diff
@@ -441,23 +458,25 @@ const CODE_CALLER_CONTRACT: &str = r#"
 	let (caller_wasm, caller_code_hash) = compile_module::<Test>(CODE_CALLER_CONTRACT).unwrap();
-	let verifiable_doughnut = MockDoughnut::new(true);
+	let verifiable_doughnut = MockDoughnut::default()
+		.set_runtime_verifiable(true)
```
Too good 🔥

jordy25519 reviewed (approved):
🚀 Looks good!
